src/isolate.rs

```diff
@@ -857,25 +857,129 @@ impl IsolateHandle {
   }
 }
 
+#[derive(Debug)]
+#[repr(C)]
+/// v8::Locker is a scoped lock object. While it's active, i.e. between its
+/// construction and destruction, the current thread is allowed to use the locked
+/// isolate. V8 guarantees that an isolate can be locked by at most one thread at
+/// any time. In other words, the scope of a v8::Locker is a critical section.
+pub struct Locker {
+  has_lock: bool,
+  top_level: bool,
+  isolate: *mut Isolate,
+}
+
+impl Locker {
+  pub fn new(isolate: *mut Isolate) -> Self {
+    Self {
+      has_lock: false,
+      top_level: true,
+      isolate,
+    }
+  }
+  pub fn unlock(&mut self) {
+    self.has_lock = false;
+  }
+  pub fn lock(&mut self) {
+    self.has_lock = true;
+  }
+}
+
+#[derive(Debug)]
+#[repr(C)]
+pub struct Unlocker {
+  isolate: *mut Isolate,
+}
+
+impl Unlocker {
+  pub fn new(isolate: *mut Isolate) -> Self {
+    Self {
+      isolate,
+    }
+  }
+}
+
+impl Drop for Locker {
+  fn drop(&mut self) {
+    unsafe {
+      println!("dropped the locker");
+      let iso = self.isolate.as_mut().unwrap();
+      iso.exit();
+      // iso.dispose();
+    }
+  }
+}
+
+extern "C" {
+  fn v8__Locker__CONSTRUCT(buf: *mut Locker, isolate: *mut Isolate);
+  fn v8__Locker__DESTRUCT(this: &mut Locker);
+  fn v8__Unlocker__CONSTRUCT(buf: *mut Unlocker, isolate: *mut Isolate);
+  fn v8__Unlocker__DESTRUCT(this: &mut Unlocker);
+}
+
+pub struct UnlockedIsolate {
+  cxx_isolate: NonNull<Isolate>,
+  locker: Locker,
+  unlocker: Unlocker,
+}
```

Comment:
We may want to merge the concept of `UnlockedIsolate` with `IsolateHandle` to simplify the API and they have very related purposes.